### PR TITLE
[QDP] Add a Quantum Data Loader and API refactor

### PR DESCRIPTION
**Pull request by rich7420**
### Purpose of PR

Introduces a **stateful Rust iterator** (Quantum Data Loader) that yields one encoded batch (DLPack) per step, so Python can drive encoding with `for qt in loader:` instead of a closed benchmark loop. The same Rust core supports both **benchmark** (full pipeline, stats only) and **data loader** (batch-by-batch iteration). The public Python API is moved out of `benchmark/` into a package **`qumat_qdp`**; benchmark scripts import from `qumat_qdp` and remain the primary consumers.

### qdp-core (Rust)

- **`pipeline_runner.rs`** (new): `PipelineConfig`, `PipelineRunResult`, `run_throughput_pipeline`, `run_latency_pipeline`; shared `vector_len`, `generate_batch`, `fill_sample`. Adds **`DataSource`** enum (Synthetic only; File reserved for Phase 2) and **`PipelineIterator`** with `new_synthetic(engine, config)` and `next_batch(&mut self) -> Result<Option<*mut DLManagedTensor>>` reusing `encode_batch`.
- **`lib.rs`**: `QdpEngine` implements **`Clone`**; re-exports `DataSource`, `PipelineIterator`, `PipelineConfig`, `PipelineRunResult`, `run_throughput_pipeline`, `run_latency_pipeline` (Linux only).
- **`gpu/encodings/amplitude.rs`**: `run_amplitude_dual_stream_pipeline` and exposure for dual-stream encode path used by benchmarks.

### qdp-python (Rust bindings)

- **`SendPtr`** wrapper: wraps `*mut DLManagedTensor` so the raw pointer can cross `py.allow_threads` (closure return must be `Send`). Used only to release GIL during encode.
- **`PyQuantumLoader`** (`#[pyclass]`): holds `Option<PipelineIterator>`. Implements Python iterator protocol: `__iter__` returns `self`; `__next__` takes iterator out with `take()`, runs `next_batch()` inside `py.allow_threads`, then restores the iterator or returns a `QuantumTensor`. Raises `StopIteration` when exhausted. Stub on non-Linux.
- **`QdpEngine::create_synthetic_loader(...)`**: builds `PipelineConfig`, calls `PipelineIterator::new_synthetic(engine.clone(), config)`, returns `PyQuantumLoader`. Stub on non-Linux.
- **`run_throughput_pipeline_py`**: pipeline run is executed inside **`py.allow_threads`** so the full benchmark loop runs with GIL released (replacing previous detach/allow_threads usage as appropriate).

### qumat_qdp (Python package)

- **`qumat_qdp/`** at project root: `__init__.py` (re-exports `QdpEngine`, `QuantumTensor`, `QdpBenchmark`, `ThroughputResult`, `LatencyResult`, `QuantumDataLoader`, `run_throughput_pipeline_py`), **`api.py`** (QdpBenchmark, ThroughputResult, LatencyResult; calls `_qdp.run_throughput_pipeline_py`), **`loader.py`** (QuantumDataLoader builder; `__iter__` calls `engine.create_synthetic_loader` and returns the Rust iterator).
- **`pyproject.toml`**: **`python-source = "."`** so the root `qumat_qdp` package is included in the wheel.

### Benchmark scripts

- **Imports**: All benchmark scripts now use `from qumat_qdp import ...` (or re-export). **`benchmark/api.py`** and **`benchmark/loader.py`** only re-export from `qumat_qdp` for backward compatibility.
- **Import path**: Each script inserts the project root into `sys.path` before importing `qumat_qdp`, so `uv run python benchmark/run_pipeline_baseline.py` (and similar) works without extra `PYTHONPATH`.
- **`benchmark_loader_throughput.py`** (new): runs throughput by iterating `QuantumDataLoader` (`for qt in loader`) and compares with `QdpBenchmark.run_throughput()` (full Rust pipeline).

## Behaviour

- **Benchmark path**: Unchanged from caller’s perspective. `QdpBenchmark(device_id=0).qubits(16).batches(100, 64).run_throughput()` still runs the full pipeline in Rust with GIL released via `run_throughput_pipeline_py`.
- **Loader path**: `QuantumDataLoader(device_id=0).qubits(16).batches(100, 64).source_synthetic()` then `for qt in loader:` yields one `QuantumTensor` (batch) per iteration; GIL is released during each `next_batch()` in Rust.

## Testing

- Existing benchmark flow: `uv run python benchmark/run_pipeline_baseline.py`, `benchmark_throughput.py`, `benchmark_latency.py`.
- Loader throughput: `uv run python benchmark/benchmark_loader_throughput.py` (compares loader iteration vs full pipeline).

### Related Issues or PRs

Related to #969 

### Changes Made

- [ ] Bug fix
- [x] New feature
- [x] Refactoring
- [ ] Documentation
- [ ] Test
- [ ] CI/CD pipeline
- [ ] Other

### Breaking Changes

- [x] Yes
- [ ] No

### Checklist

- [x] Added or updated unit tests for all changes
- [x] Added or updated documentation for all changes
- [x] Successfully built and ran all unit tests or manual tests locally
- [ ] PR title follows "MAHOUT-XXX: Brief Description" format (if related to an issue)
- [x] Code follows ASF guidelines

### CODE REVIEW

**Comment by rich7420** [comment]
Before
Running throughput: 5 trials (qubits=16, batch_size=64, prefetch=16, batches=200)
Throughput: median=1153.5 vec/s, p95=1279.0 vec/s
Running latency: 5 trials
Latency: median=0.853 ms/vec, p95=0.874 ms/vec

After
Running throughput: 5 trials (qubits=16, batch_size=64, prefetch=16, batches=200)
Throughput: median=2541.9 vec/s, p95=3000.6 vec/s
Running latency: 5 trials
Latency: median=0.373 ms/vec, p95=0.401 ms/vec

**Comment by rich7420** [comment]
cc @viiccwen

**Comment by ryankert01** [comment]
Love the idea! When will 0.6.0 release dates be? I think we should land it faster.

> Btw, can you share the benchmark result? Do it increase speed for decrease speed? I might be able to help with that.
edit: sorry I didn't saw it at first place!

**Comment by guan404ming** [comment]
We could prepare it recently I think. We could iterate it fast.

File: qdp/qdp-python/benchmark/benchmark_latency.py [hunk 1]
```diff
@@ -26,12 +26,20 @@
...
 import argparse
+import sys
 import time
+from pathlib import Path
 
 import torch
 
-from _qdp import QdpEngine
-from utils import normalize_batch, prefetched_batches
+# Add project root to path so qumat_qdp is importable when run as script
```
Comment (guan404ming) [inline qdp/qdp-python/benchmark/benchmark_latency.py]: I think we could maybe handle this more better by using the internal qdp export here?

File: qdp/qdp-python/benchmark/benchmark_loader_throughput.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+"""
+
+from __future__ import annotations
+
+import argparse
+import sys
+import time
+from pathlib import Path
+
+# Add project root to path so qumat_qdp is importable when run as script
```
Comment (guan404ming) [inline qdp/qdp-python/benchmark/benchmark_loader_throughput.py]: ditto

File: qdp/qdp-python/benchmark/benchmark_throughput.py [hunk 3]
```diff
@@ -28,13 +28,21 @@
...
+import sys
 import time
+from pathlib import Path
 
 import numpy as np
 import torch
 
-from _qdp import QdpEngine
-from utils import normalize_batch, prefetched_batches
+# Add project root to path so qumat_qdp is importable when run as script
```
Comment (guan404ming) [inline qdp/qdp-python/benchmark/benchmark_throughput.py]: ditto

File: qdp/qdp-python/qumat_qdp/__init__.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    from qumat_qdp import QdpBenchmark, ThroughputResult, LatencyResult
+    from qumat_qdp import QuantumDataLoader
+"""
+
+from __future__ import annotations
+
+# Rust extension (built by maturin)
+import _qdp
+
+from qumat_qdp.api import (
```
Comment (guan404ming) [inline qdp/qdp-python/qumat_qdp/__init__.py:36]: Do we want to export these apis to our users?

File: qdp/qdp-python/qumat_qdp/loader.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+from __future__ import annotations
+
+from typing import Iterator, Optional
+
+# Lazy import _qdp until __iter__ is used
+_qdp: Optional[object] = None
+
+
+def _get_qdp():
```
Comment (guan404ming) [inline qdp/qdp-python/qumat_qdp/loader.py:47]: I'm not really sure if we need this. Could you explain more?

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -71,7 +71,7 @@ repos:
         name: cargo clippy
         language: system
         pass_filenames: false
-        entry: bash -c 'LIBTORCH_USE_PYTORCH=1 cargo clippy "$@"' --
+        entry: bash -c 'LIBTORCH_USE_PYTORCH=1 LIBTORCH_BYPASS_VERSION_CHECK=1 cargo clippy "$@"' --
```
Comment (guan404ming) [inline .pre-commit-config.yaml]: I think we use local torch instead of libtorch here. Could you help share why we need this change?

Comment (rich7420) [reply .pre-commit-config.yaml]: oh , cuz my local pre-commit errors.
I could remove this part.

File: qdp/qdp-python/benchmark/benchmark_latency.py (hunk 1 above)
Comment (rich7420) [reply qdp/qdp-python/benchmark/benchmark_latency.py]: Done. I removed the `sys.path` hack and rely on the package being importable when run from the qdp-python directory.

File: qdp/qdp-python/qumat_qdp/__init__.py (hunk 4 above)
Comment (rich7420) [reply qdp/qdp-python/qumat_qdp/__init__.py:36]: I think we should intend to expose `QdpEngine` and `QuantumTensor` (and optionally `run_throughput_pipeline_py`) for advanced users who need direct control; `QdpBenchmark` and `QuantumDataLoader` remain the recommended high-level API. WDYT?

File: qdp/qdp-python/qumat_qdp/loader.py (hunk 5 above)
Comment (rich7420) [reply qdp/qdp-python/qumat_qdp/loader.py:47]: yeah , you're right, we don't need it in this part

File: qdp/qdp-python/benchmark/benchmark_loader_throughput.py (hunk 2 above)
Comment (rich7420) [reply qdp/qdp-python/benchmark/benchmark_loader_throughput.py]: done

File: qdp/qdp-python/benchmark/benchmark_throughput.py (hunk 3 above)
Comment (rich7420) [reply qdp/qdp-python/benchmark/benchmark_throughput.py]: done

File: qdp/qdp-python/qumat_qdp/loader.py (hunk 5 above)
Comment (rich7420) [reply qdp/qdp-python/qumat_qdp/loader.py:47]: I had intended the lazy import to defer loading the Rust extension until first use, but with the current design it doesn’t actually achieve that

File: qdp/qdp-python/qumat_qdp/__init__.py (hunk 4 above)
Comment (guan404ming) [reply qdp/qdp-python/qumat_qdp/__init__.py:36]: Make sense, thanks!

File: qdp/qdp-python/benchmark/benchmark_latency.py (hunk 1 above)
Comment (guan404ming) [reply qdp/qdp-python/benchmark/benchmark_latency.py]: Thanks!

**Comment by guan404ming** [comment]
Please help resolve ci error thanks!

File: qdp/qdp-python/qumat_qdp/loader.py (hunk 5 above)
Comment (guan404ming) [reply qdp/qdp-python/qumat_qdp/loader.py:47]: After some investigation, I found this piece of code seems work for this issue. Could you help test it?

```
from functools import lru_cache

@lru_cache(maxsize=1)
def get_qdp():
    import _qdp
    return _qdp
```

Comment (guan404ming) [reply qdp/qdp-python/qumat_qdp/loader.py:47]: I tested it seems not work. Maybe we could also try to move the import `_qdp` to `if TYPE_CHECKING:` and rewrite the type def to `def func(x: "_qdp.Type"):` and see if it works

Comment (rich7420) [reply qdp/qdp-python/qumat_qdp/loader.py:47]: got it! thanks for the advice!

**Comment by viiccwen** [comment]
It's failed with testing (CI didn't trigger these tests due to lack of CUDA environment).

```
testing/qdp/test_bindings.py::test_encode_cuda_tensor[data_shape0-expected_shape0-1] FAILED                                                                [  3%]
testing/qdp/test_bindings.py::test_encode_cuda_tensor[data_shape1-expected_shape1-3] FAILED                                                                [  3%]
testing/qdp/test_bindings.py::test_encode_cuda_tensor_preserves_input[data_shape0-False] FAILED                                                            [  3%]
testing/qdp/test_bindings.py::test_encode_cuda_tensor_preserves_input[data_shape1-True] FAILED                                                             [  4%]
testing/qdp/test_bindings.py::test_encode_3d_rejected[cuda_tensor-Unsupported CUDA tensor shape: 3D] FAILED                                                [  4%]
testing/qdp/test_bindings.py::test_encode_cuda_tensor_non_finite_values[<lambda>-zeros] FAILED                                                             [  4%]
testing/qdp/test_bindings.py::test_encode_cuda_tensor_non_finite_values[<lambda>-NaN] FAILED                                                               [  5%]
testing/qdp/test_bindings.py::test_encode_cuda_tensor_non_finite_values[<lambda>-Inf] FAILED                                                               [  5%]
testing/qdp/test_bindings.py::test_encode_cuda_tensor_output_dtype[float32-expected_dtype0] FAILED                                                         [  5%]
testing/qdp/test_bindings.py::test_encode_cuda_tensor_output_dtype[float64-expected_dtype1] FAILED                                                         [  5%]
```

File: qdp/qdp-python/src/lib.rs
Comment (viiccwen) [inline qdp/qdp-python/src/lib.rs]: We don't need this since PyTorch default stream can report cuda_stream as 0.
I don't know whether ur change cause this or not. Just found.
Then we can pass the testing. : D

File: qdp/qdp-python/src/lib.rs [hunk 7]
```diff
@@ -1038,20 +1112,193 @@ impl QdpEngine {
...
+    ) -> PyResult<PyQuantumLoader> {
+        let config = PipelineConfig {
+            device_id: 0,
+            num_qubits,
+            batch_size,
+            total_batches,
+            encoding_method: encoding_method.to_string(),
+            seed,
+            warmup_batches: 0,
+        };
```
Comment (viiccwen) [inline qdp/qdp-python/src/lib.rs:1158]: device_id uses hard-coded 0, config should equals to use actually.

File: testing/qdp/test_benchmark_api.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    assert result.latency_ms_per_vector > 0
+
+
+@requires_qdp
+def test_qdp_benchmark_validation():
+    """QdpBenchmark.run_throughput() raises if qubits/batches not set."""
+    import api
+
+    with pytest.raises(ValueError, match="qubits and batches"):
+        api.QdpBenchmark(device_id=0).run_throughput()
```
Comment (viiccwen) [inline testing/qdp/test_benchmark_api.py]: lacks of testing of `run_latency()` if qubits/batches not set.

File: qdp/qdp-python/qumat_qdp/loader.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    """
+
+    def __init__(
+        self,
+        device_id: int = 0,
+        num_qubits: int = 16,
+        batch_size: int = 64,
+        total_batches: int = 100,
+        encoding_method: str = "amplitude",
+        seed: Optional[int] = None,
```
Comment (viiccwen) [inline qdp/qdp-python/qumat_qdp/loader.py:103]: seed in Rust interface is: `seed: Option<u64>`, but in Python is: `seed: Optional[int] = None`.

maybe we can add validate guard to raise `ValueError` with clear description.

File: testing/qdp/test_benchmark_api.py (hunk 8 above)
Comment (ryankert01) [reply testing/qdp/test_benchmark_api.py]: In test we also don't have `device_id`propagation.

File: qdp/qdp-python/qumat_qdp/loader.py (hunk 9 above)
Comment (ryankert01) [reply qdp/qdp-python/qumat_qdp/loader.py:103]: Validate every args user enters and raise proper error might be a good one.

Comment (rich7420) [reply qdp/qdp-python/qumat_qdp/loader.py:103]: updated! plz take a another look

**Comment by ryankert01** [comment]
Overall lgtm

**Comment by rich7420** [comment]
thanks for the review @guan404ming , @ryankert01 and @viiccwen !
